### ISSUE TITLE
Type mismatch of the right side of a statement involving the ?? operator is ignored if the left side is of type `dynamic`

### ISSUE THREAD

**Issue by vxern** [dart-lang/linter#2781]
**Describe the issue**
The linter does not catch a type mismatch between the right side of a statement and the variable being assigned with the null-aware ?? operator if the left side is `dynamic`.

**To Reproduce**
```dart
dynamic possiblyNull;
Map<String, dynamic> clearlyAMap = {};
clearlyAMap = possiblyNull ?? 'this is clearly a string';
```

**Expected behavior**
The variable on the right should be flagged up as having a different type to the variable being assigned.

**Additional context**
None

**Comment by bwilkerson** [comment]
@eernstg

**Comment by eernstg** [comment]
This is working as intended. `clearlyAMap` has type `Map<String, dynamic>`, so anything assigned to that variable must have a static type which is assignable to that. But `dynamic` is assignable to anything (and it causes a dynamic check to be generated, such that we verify at run time that the assigned object is _actuall_ a `Map<String, dynamic>`).

So we just need to consider why `possiblyNull ?? 'this is clearly a string'` has type `dynamic`: That's because the type `T` of `e1 ?? e2` is `UP(NON_NULL(T1), T2)` where `T1` is the type of `e1`, `T2` is the type of `e2`, `UP` is the [least upper bound function](https://github.com/dart-lang/language/blob/master/resources/type-system/upper-lower-bounds.md#upper-bounds), and `NON_NULL` is the function that removes `?` from a type, if possible. We don't have a mechanism that eliminates `Null` from an arbitrary type, so it only changes types where the nullability is explicit: `NON_NULL(int?)` is `int` and `NON_NULL(X?)` is `X`, but `NON_NULL(dynamic)` is `dynamic`. This implies that `T` is `dynamic`.

**Comment by vxern** [comment]
> This is working as intended. `clearlyAMap` has type `Map<String, dynamic>`, so anything assigned to that variable must have a static type which is assignable to that. But `dynamic` is assignable to anything (and it causes a dynamic check to be generated, such that we verify at run time that the assigned object is _actuall_ a `Map<String, dynamic>`).
> 
> So we just need to consider why `possiblyNull ?? 'this is clearly a string'` has type `dynamic`: That's because the type `T` of `e1 ?? e2` is `UP(NON_NULL(T1), T2)` where `T1` is the type of `e1`, `T2` is the type of `e2`, `UP` is the [least upper bound function](https://github.com/dart-lang/language/blob/master/resources/type-system/upper-lower-bounds.md#upper-bounds), and `NON_NULL` is the function that removes `?` from a type, if possible. We don't have a mechanism that eliminates `Null` from an arbitrary type, so it only changes types where the nullability is explicit: `NON_NULL(int?)` is `int` and `NON_NULL(X?)` is `X`, but `NON_NULL(dynamic)` is `dynamic`. This implies that `T` is `dynamic`.
> 
> @vxern, you might prefer to use an options file with the analyzer specifying the following:
> 
> ```
> analyzer:
>   strong-mode:
>     implicit-casts: false
> ```
> 
> This will make the analyzer flag a cast from dynamic to anything other than a top type.

Thank you for the great reply. I understand now 😁
